### DISCUSSION TITLE
Stops working after long gap with no speech?

### DISCUSSION THREAD

**Discussion by gmaxwell** [openai/whisper#29] [Q&A] [answered]
I have some inputs in  Norwegian with 50 minute long segments with no speech and only background noise (a lunchbreak).  Whisper falls into a loop repeating the last(?) segment spoken and then after the speech resumes it never breaks out of it. Same behavior in translate and transcribe mode (though the output it repeats is different). 

Truncating off the non-speech segment and restarting gets me a sensible transcription.

I can put up a test file if it would be useful, though I assume the behavior is easily reproducible.

**Comment by jonhassall** [comment]
I get a similar behavior with no speech and only background noise periods in English and Spanish audio. Sometimes it will make comical nonsensical transcriptions.

One quiet period produced 'I hope you liked the video, don't forget to leave a like and subscribe to the channel, thank you very much for your attention and see you in the next video,' despite this never being said.

Interesting how humans can hear things when too much silence too.

**Reply by olastor** [reply]
For me it transcribed "[03:05.120 --> 03:10.680]  Find out more at www.beadaholique.com to purchase beading supplies and to get design ideas!" in a longer section with only silence :smile:

**Reply by Shredsauce** [reply]
> For me it transcribed "[03:05.120 --> 03:10.680] Find out more at [www.beadaholique.com](http://www.beadaholique.com) to purchase beading supplies and to get design ideas!" in a longer section with only silence 😄

This same thing happened to me and you're the only other one I've found who's seen it too. That website probably managed to sneak a spammy commit into the repo. For now some kind of regex thing to filter out spam will have to do.

**Reply by endimionzf** [reply]
I get the same like subscribe message, but in romanian

**Comment by shirayu** [comment]
I got the same problem.

I think predicted `no_speech_prob` values are too small for silent periods.

No update of `prompt_reset_since` may also cause repeating outputs.
If the following code is not executed, 
https://github.com/openai/whisper/blob/8cf36f3508c9acd341a45eb2364239a3d81458b9/whisper/transcribe.py#L220-L222
past tokens are used as prompt.
https://github.com/openai/whisper/blob/8cf36f3508c9acd341a45eb2364239a3d81458b9/whisper/transcribe.py#L162

Past strange outputs may have caused more strange outputs.

**Reply by koka0630** [reply]
your answer is helpful for me, thanks 👍 
then, how can we fix this bug in CLI? or we have to do data process by trimming silence?

**Reply by koka0630** [reply]
> how can we fix this bug in CLI? 

ah, can we decrease `no_speech_threshold`?

**Reply by shirayu** [reply]
By setting
1.  `--no_speech_threshold` more smaller  value (defualt 0.6)
2. `--logprob_threshold None` or `--logprob_threshold {small value}` (default `-1.0`)

`should_skip` become `True` and the `continue` line is executed.


https://github.com/openai/whisper/blob/15ab54826343c27cfaf44ce31e9c8fb63d0aa775/whisper/transcribe.py#L251-L252

https://github.com/openai/whisper/blob/15ab54826343c27cfaf44ce31e9c8fb63d0aa775/whisper/transcribe.py#L166-L175

This has the effect of not transcribing silence intervals.
However, note that if the predicted `logprob_threshold` value is low for the voices, they will not be transcribed.

For more substantive improvements, we need to make the model improve to output `no_speech_threshold` more accurately and set appropriate thresholds.

**Reply by shirayu** [reply]
According to [their paper](https://cdn.openai.com/papers/whisper.pdf), they used 30-second slient segments for predicting `<|nospeech|>`.

> We break audio files into 30-second segments paired with the subset of the transcript that occurs within that time segment. We train on all audio, including segments where there is no speech (though with sub-sampled probability) and use these segments as training data for voice activity detection.
>  In the case where there is no speech in an audio segment, the model is trained to predict a <|nospeech|> token indicating this.

I am interested in the amount and percentage of "silence", what kind of "silence" is included, and the performance of prediction of "silence" (i.e. VAD = Voice activity detection).

**Comment by jongwook** [comment, marked as answer]
This is one of the limitations of the current hacky approach to long-form transcription. The VAD output from the model is not very accurate, and the predicted `no_speech_prob` is often not a reliable predictor of voice activity.

I chose the default `no_speech_threshold` of 0.6 which worked okay for a few datasets that I tested with, but different combinations of `--compression_ratio_threshold`, `--logprob_threshold`, and `--no_speech_threshold` values might be needed depending on the audio.

There's also a hard-coded constant as @shirayu mentioned, which determines whether the text from the previous window gets fed as the prompt:

https://github.com/openai/whisper/blob/2d3032de014c514d5c029baccb898b5342463674/whisper/transcribe.py#L220-L222

This is supposed to help the text flow more naturally between window boundaries, but as @shirayu mentioned, this makes the decoding more prone to repetition looping.

I'll plan to document this limitation more clearly in README.md.

**Reply by shirayu** [reply]
Thank you for your answer!
I will think about using another VAD method together in my streaming application [(whispering)](https://github.com/shirayu/whispering) .

**Reply by light42** [reply]
How about using [webrtcvad](https://github.com/wiseman/py-webrtcvad)? It is quite outdated and has some limits, but when I used it in my project it produced alright results.

**Reply by AdolfVonKleist** [reply]
https://github.com/snakers4/silero-vad is _much_ better than webrtcvad IMO.  Quite easy to use and worth a look.

**Reply by shirayu** [reply]
@AdolfVonKleist Thank you for the nice link.
I made [whispering](https://github.com/shirayu/whispering) drop every 3.75 sec silent period as default by using it.

**Reply by cndhng** [reply]
Hi, I am new to CLI and still not used to some of these parameters. Would you mind explaining
 `--no_speech_threshold` `--compression_ratio_threshold` `--logprob_threshold` and how you write them syntactically correct in the cmd prompt, I have a similar issue where there are some longer no-speech periods and it always uses the last detected speech segment and repeats through the end of the transcription.

**Reply by StefanNa3Shape** [reply]
So I have the same issue and I used pyannotes VAD https://github.com/pyannote/pyannote-audio/blob/develop/tutorials/voice_activity_detection.ipynb 
That works very well for me in terms of finding the segments with voice.
But then I run whisper on all the segments with speech and it still hallucinates a lot.
I lowered no_speech_threshold but that does not make a difference.
Does anyone have new insights?

**Comment by ANonEntity** [comment]
I've managed to mostly solve this by segmenting the input audio with Silero VAD. Here's my implementation:
https://colab.research.google.com/github/ANonEntity/WhisperWithVAD/blob/main/WhisperWithVAD.ipynb

**Reply by ejkitchen** [reply]
Thanks for this!!!

**Reply by octimot** [reply]
`# Run Whisper on each audio chunk`

Does splitting the audio into multiple chunks and feeding them individually to Whisper degrade the quality of the final result? 

I noticed that running Whisper on longer audio files gets way better translations to english, compared to using shorter ones.

**Reply by ANonEntity** [reply]
It improves it, from my testing. Whisper outputs hallucinations during long periods of silence, which is when the notebook splits the audio. If the pause is shorter than 3 seconds, it'll just remove the silent bit without splitting the file.

**Reply by maphew** [reply]
> Does splitting the audio into multiple chunks and feeding them individually to Whisper degrade the quality of the final result?
> 
> I noticed that running Whisper on longer audio files gets way better translations to english, compared to using shorter ones.

Do have any numbers on the splits? e.g. 30min is better than 15min. but worse than 60min?

**Reply by octimot** [reply]
@ANonEntity I think I didn't phrase the question correctly. My concern wasn't about deleting the silence (that's cool!), but about running whisper on each audio chunk versus re-merging the audio (without the periods of silence) and feeding a single file to Whisper. 

In other words, wouldn't Whisper give out better results if you feed it a long audio (without silence) instead of feeding it the split audio files with the same content? 

@maphew I was only hypothesising based on a few tests but hardly anything relevant... For eg. a 30 second video in Chinese produced less reliable translations (and different ones on each run), while a 30 minute video in Spanish produced very accurate translations. 

Testing same language and equal quality sources would probably be more relevant in reaching a conclusion.

**Reply by octimot** [reply]
To add to the previous reply:

I have just transcribed (translated) an audio file which is a recording of Person A telling phrases in English to Person B, which Person B repeats in Spanish. The audio is approx. 20 minutes in total.

For the first 3 minutes, Whisper translates the Spanish phrases to English, but after that it stops translating them, almost as if it recognized that Person B repeats in Spanish exactly what Person A is saying. Is this possible?

I guess that this is what I was referring to when I posed my question regarding using chunks versus one single audio file. It feels like Whisper gets better in handling the audio depending on how much of the context it understands. Or is this all just accidental?

**Reply by light42** [reply]
Whisper default VAD is better than this when tested using audio which has constant background noise. The most obvious example would be songs.

**Reply by alexlyzhov** [reply]
Thank you! Works like magic for my files with very long silence periods.

**Reply by aadnk** [reply]
Using a VAD seems to be the key to getting a good transcript for very long files, so thanks for the suggestion. 👍 

I did something [similar](https://huggingface.co/spaces/aadnk/whisper-webui/blob/main/src/vad.py#L83) in my [WebUI implementation](https://huggingface.co/spaces/aadnk/whisper-webui) of Whisper (though by default, I merge speech sections into 150 s chunks, at most).

I've found that this method can handle even video files as [long as 7-8 hours](https://www.reddit.com/r/LearnJapanese/comments/xljc3e/whisper_a_new_free_ai_model_from_openai_that_can/irs7qos/), though on occasion (once an hour) there will be a section with nonsensical transcription. A secondary benefit is that the timings of each timestamp no longer goes out or sync - or rather, that it is readjusted every time whisper is re-run on a new section of speech.

**Reply by KTRosenberg** [reply]
> Using a VAD seems to be the key to getting a good transcript for very long files, so thanks for the suggestion. 👍
> 
> I did something [similar](https://huggingface.co/spaces/aadnk/whisper-webui/blob/main/src/vad.py#L83) in my [WebUI implementation](https://huggingface.co/spaces/aadnk/whisper-webui) of Whisper (though by default, I merge speech sections into 150 s chunks, at most).
> 
> I've found that this method can handle even video files as [long as 7-8 hours](https://www.reddit.com/r/LearnJapanese/comments/xljc3e/whisper_a_new_free_ai_model_from_openai_that_can/irs7qos/), though on occasion (once an hour) there will be a section with nonsensical transcription. A secondary benefit is that the timings of each timestamp no longer goes out or sync - or rather, that it is readjusted every time whisper is re-run on a new section of speech.

Is there a python example of this working with microphone input? I was hoping to support continuous recognition with incremental hypotheses.

**Reply by alexlyzhov** [reply]
@KTRosenberg Maybe this would work: https://github.com/shirayu/whispering. They use snakers4/silero-vad as well.

**Reply by KTRosenberg** [reply]
> @KTRosenberg Maybe this would work: https://github.com/shirayu/whispering. They use snakers4/silero-vad as well.

Ah, I'll try this. However, should I somehow integrate one of the alternative VADs discussed here? If I interpret the conversation correctly, longform transcription could lead to loops.

One thing I wonder: would it be possible to (programmatically) split the audio and restart the transcription between sentence boundaries automatically? For example, if a piece of the transcription is final? That could avoid too much memory use. A lot of the commercial solutions output continuous hypotheses as words are recognized, but force cancel the transcription after a timeout. I'm trying to have a little more control over that so words don't get cut-off. Hopefully this makes sense.

**Reply by aadnk** [reply]
> One thing I wonder: would it be possible to (programmatically) split the audio and restart the transcription between sentence boundaries automatically? For example, if a piece of the transcription is final? That could avoid too much memory use. A lot of the commercial solutions output continuous hypotheses as words are recognized, but force cancel the transcription after a timeout. I'm trying to have a little more control over that so words don't get cut-off. Hopefully this makes sense.

So if I understand you correctly, you'd like to run Whisper on the microphone input as soon as possible, generating some guesses. But then re-run Whisper on the whole sentence, once the VAD (for instance snakers4/silero) has detected a sentence boundary/non-speech part? 

I haven't checked if this isn't already done by [shirayu/whispering](https://github.com/shirayu/whispering), but it definitely sounds possible. For instance, you could have two buffers that is continuously feed by the microphone asynchronously - one that is processed nearly instantly for continuous hypotheses, and one that's only processed once the VAD detects lack of speech for some minimum duration. You can also pass the [previous detected text as a prompt](https://github.com/openai/whisper/blob/main/whisper/decoding.py#L87) to Whisper, to improve the detection in each speech chunk.

And of course, the GPU has to be powerful enough to keep up with the input from the microphone - though of course, if that isn't always the case, you could let the buffer for continuous hypotheses fill up and just not generate hypotheses while you still generate full sentences from the other buffer.

**Reply by KTRosenberg** [reply]
> > One thing I wonder: would it be possible to (programmatically) split the audio and restart the transcription between sentence boundaries automatically? For example, if a piece of the transcription is final? That could avoid too much memory use. A lot of the commercial solutions output continuous hypotheses as words are recognized, but force cancel the transcription after a timeout. I'm trying to have a little more control over that so words don't get cut-off. Hopefully this makes sense.
> 
> So if I understand you correctly, you'd like to run Whisper on the microphone input as soon as possible, generating some guesses. But then re-run Whisper on the whole sentence, once the VAD (for instance snakers4/silero) has detected a sentence boundary/non-speech part?
> 
> I haven't checked if this isn't already done by [shirayu/whispering](https://github.com/shirayu/whispering), but it definitely sounds possible. For instance, you could have two buffers that is continuously feed by the microphone asynchronously - one that is processed nearly instantly for continuous hypotheses, and one that's only processed once the VAD detects lack of speech for some minimum duration. You can also pass the [previous detected text as a prompt](https://github.com/openai/whisper/blob/main/whisper/decoding.py#L87) to Whisper, to improve the detection in each speech chunk.
> 
> And of course, the GPU has to be powerful enough to keep up with the input from the microphone - though of course, if that isn't always the case, you could let the buffer for continuous hypotheses fill up and just not generate hypotheses while you still generate full sentences from the other buffer.

I think this is correct, yes.Basically, a previous commercial solution I was using would fire callbacks upon a new hypothesis/word detected, and timeout after a fixed amount of silence. The difficulty there was figuring out how to cross the "boundary" between different chunks. If Whisper could make this easier, that would be great. I could probably pass-in some fixed window of previous text to keep things bounded.

Anyway, I think your description makes sense, though I'm still exploring the API. Besides shirayu's repository, are there examples of creating this sort of multi-buffer solution? I'm not actually aware of how one would process data asynchronously in Python since as I recall, it's always one thread.

"one that's only processed once the VAD detects lack of speech for some minimum duration" -- I think I would just set a fixed timeout since my understanding from this discussion is that VAD could break.

**Reply by phineas-pta** [reply]
I adapted your notebook to Faster Whisper, with some minor adjustments for Japanese: https://github.com/phineas-pta/VwArship

**Reply by tvone** [reply]
@phineas-pta not working

**Comment by gmaxwell** [comment]
FWIW, I've found that the current code (vs the day  1 release) works much better on my inputs and is now able to transcribe complete recordings that the original code choked on.

**Reply by alexlyzhov** [reply]
I tried the current code and it still has the same behavior for me in terms of handling long silences (can't transcribe, endless repetitions/hallucinations)

**Comment by nicholasgcotton** [comment]
I have the same issue with multilingual audio files on the CLI version of whisper, if there is any significant silence whisper does not resume transcribing and instead repeats the last thing said forever. For whatever reason, this does not occur if I call Whisper from within a python program.

To get around this I created a little program to call the python version of Whisper from the command line. I know next to nothing about python so there's very likely an easier option, but in the meantime this does work. I call it WhisperDO.

WhisperDO.py:

```
import whisper
from whisper.transcribe import cli
cli()
```
It can be called as **python whisperDO.py --args** and accepts all the same inputs as the CLI version (as far as I can tell).

**Reply by ejkitchen** [reply]
What parameters are you passing to Whisper?

**Reply by nicholasgcotton** [reply]
Just the basics, --language, --task, and --model. 

I tried changing compression ratio a number of times but it had no effect as far as I could tell. Issue has repeated with"Chinese" and Punjabi conversations, although in both cases the audio started with an English introduction. Removing silence from the files below -20bd using Audacity also fixes the issue, but then of course SRT/VTT times are off.

**Reply by nicholasgcotton** [reply]
I'm limited to the small and medium models as I've currently only got an T1200 and RTX 3070 available for testing.

**Comment by Shamshiel** [comment]
I wonder is that an issue that can be fixed in Whisper itself or do we have to resort to use something like Silero VAD?

**Comment by eliemaalouly** [comment]
You can try this script: [whisper-vad](https://github.com/e-maalouly/Transcription-whisper_pyannote).
It uses a VAD to detect voice activity, and only feeds those segments to Whisper. It fixed the loop issues for me.

**Reply by cndhng** [reply]
Hello, would you mind to show me how to use it with whisper. I'm new to this how CLI thing. Thank you

**Reply by turnkit** [reply]
Does this change the timing of the source file vs. the timed subtitles?  I need the output to match the original.

Does using this method require the audio to be uploaded to a service, with an API it would appear so.  I need to keep all my data only locally. Could there be a solution to get it to run locally?

**Comment by MatthiasReumann** [comment]
For those who wonder how one can adjust `--no_speech_threshold` and `--logprob_threshold` (or any other parameter actually) in Python using the `DecodingOptions` object:
```python
model = whisper.load_model('tiny')
options = whisper.DecodingOptions().__dict__.copy()
options['no_speech_threshold'] = 0.275
options['logprob_threshold'] = None

result = model.transcribe('/path/to/file', **options, verbose=False)
```

A full example in action can be seen at: [whisper_transcriber.py](https://github.com/TUM-Dev/TUM-Live-Voice-Service/blob/main/subtitles/whisper_transcriber.py).

**Comment by BJKriptal** [comment]
I've found that if one adds `--condition_on_previous_text False` when starting whisper, the strange output repeating continuously will not appear.

The `--condition_on_previous_text` has this effect.
>  if True, provide the previous output of the model as a prompt for the next window; disabling may make the text inconsistent across 
> windows,` but the model becomes less prone to getting stuck in a failure loop

This works is because in `transcribe.py`, setting `condition_on_previous_text` to `False` force the following code to be executed.
```python
 if not condition_on_previous_text or result.temperature > 0.5:
                # do not feed the prompt tokens if a high temperature was used
                prompt_reset_since = len(all_tokens)
```
then the prompt_reset_since will make the next text generation not be affected by the former text.

**Comment by turnkit** [comment]
The "silence equals hallucinations" issue is a very common and very problematic issue.  

And yes, I had  --condition_on_previous_text False  as an option too.

Look at the example below!

![image](https://user-images.githubusercontent.com/4126167/229871442-925f23f7-384a-4569-ac32-ee3263395496.png)

**Reply by Ko4ka** [reply]
I think I was able to solve this one by removing segments based on their compresson ratio
https://github.com/openai/whisper/discussions/2420

**Comment by GJoe2** [comment]
What it work for me was editing the silence in an audio software (An useful tool inside audacity did the trick), obviously timestamps are messed up with the original audio, but that was not an issue for my project.

**Comment by StefanNa3Shape** [comment]
So the first thing I noticed is that if you instead use the model form Huggingface, I do not get these problems.
But with whisper from pip I do.
So try this instead:

```
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from datasets import load_dataset


device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v3"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
)

dataset = load_dataset("distil-whisper/librispeech_long", "clean", split="validation")
sample = dataset[0]["audio"]

result = pipe(sample)
print(result["text"])
```

And arguments ca be added like this:
```
generate_kwargs = {
    "max_new_tokens": 448,
    "num_beams": 1,
    "condition_on_prev_tokens": False,
    "compression_ratio_threshold": 1.35,  # zlib compression ratio threshold (in token space)
    "temperature": (0.0, 0.2, 0.4, 0.6, 0.8, 1.0),
    "logprob_threshold": -1.0,
    "no_speech_threshold": 0.6,
    "return_timestamps": True,
}

result = pipe(sample, generate_kwargs=generate_kwargs)
```


Source:
https://huggingface.co/openai/whisper-large-v3
